### ISSUE TITLE
Uphill Task

### ISSUE THREAD

**Issue by E7-87-83** [manwar/perlweeklychallenge-club#5055]
Greetings to all teammates.

**Do you want to have an extra-challenging task each month?
Or, do you accept coding extra lines in order to solve an interesting task once a month?**

Do you know an algorithm from your computer science study, would like to share with TWC teammates, but thought it might be relatively heavy?
Or, do you like to share a practical technique or procedure applied in some widely-used Perl frameworks? Why don't propose the idea as a TWC task?

Or, has an intriguing question just popped in your mind? You simply want to share it.

It is Week 135 of the TWC Calendar and 287 tasks have been appeared. Different algorithms commonly-learnt by computer science students or programmers have been used in the vast amount of solutions. Sometimes some advanced algorithms or data structures were being used. Regular participating teammates may have an expectation of the demand(difficulty, lines needed, number of test cases) of the tasks and arrange a time-slot in their planner to solve the tasks. And some techniques have been used again for the third time and you might want to code more stuff in that week.

Secretly(?), I am a distance runner. Uphill training is a demanding but beneficial workout in distance running. Since it is so demanding, most leisure runners do not put this workout in their usual schedule, or take the workout less frequent than other types of workout. However nobody declines the benefits of uphill training brings. And usually you will have a great sense of satisfaction when it's done!

So, why don't accept AN EXTRA WORKLOAD coding challenge task regularly?

---

**I suggest having an "Uphill Task 2" in the 2nd week of each month.**

It can be fun:
- create an engine to solve a traditional board game;
- code about something that professional mathematicians in number theory haven't solved. (I just come up with one such kind of coding puzzles, but not to share here in order to keep with the discussion.)

It can be practical:
- interact with a database;
- apply support for a famous framework in Perl;
- design a mini-app according to the requirements;
- something like API tasks on Week 008 to Week 024 but more sophisticated.

It can be pedagogical:
- craft a simplified interpreter;
- simulate a Turing machine;
- introduce an advanced data structure or algorithm.

**What do you think about "having an 'uphill task' in The Weekly Challenge once a month"?**
Would you think that it is better to make the "uphill task" as a "Task 3"?? (But a bit increase in administrative workload.)
Do you have any related suggestions? :)

P.S. For an example of what an "uphill task" might look like, _N Queens_ of Week 062 ([Week 062](https://theweeklychallenge.org/blog/perl-weekly-challenge-062/) , [RECAP](https://theweeklychallenge.org/blog/recap-challenge-062/)) should be a candidate. Or you get other candidates. (I still like the _[Adventure of Knight](https://theweeklychallenge.org/blog/perl-weekly-challenge-118/#TASK2)_, which I proposed, hahaha.)

-- from CY

**Comment by adamcrussell** [comment]
@E7-87-83 and I discussed this on [Discord](https://discord.gg/reQscsa2). some ideas from that discussion

- make the uphill challenges run for, say, a month
- understanding that this may add some extra administrative burden, it would be required to identify a volunteer to assemble and review the solutions, and even suggest the problems
- I'd be OK with suggesting some problems of this type to consider
- This could be a nice way to promote things such as SDL's Perl bindings for game development

@manwar Obviously none of this happens without your acceptance! I am not aware of all considerations involved, but can certainly try to help smooth anything out if I can.

**Comment by LaurentRosenfeld** [comment]
I agree with Abigail's comments. I'm not very much interested in setting up a database just for the sake of a coding challenge. I also have little interest in having to use a framework.

**Comment by Abigail** [comment]
The "problem" with something like the peg challenge is that this has already been analyzed, and you can just grab the solution from the internet. The weekly challenge is already full of challenges like this: anytime when we're asked to generate the first N numbers from an OEIS sequence for instance. 

The Adventures of Knight suffered from a similar problem: it doesn't take input, hence the program is expected to print a fixed output. And in the given problem, it was much easier to show what the optimal path was than to write a program doing the calculations. IMO, such problem aren't much of a challenge -- they're just glorified "hello world" programs. 

But it's nearly December. [Advent of Code](https://adventofcode.com/) is coming, with a challenge every day.

**Comment by adamcrussell** [comment]
@Abigail those are all fair points regarding how those specific problems might not be the best selections. I think the idea itself to have longer "marathon vs sprint" challenges has merit. To me it makes sense to just discuss that itself, and not get distracted by any specific details of the examples.

1.  Should there be "uphill" challenges? 
2. What would the first problem be?

If Mohammed agrees that the answer to (1) is yes only then proceed to examining (2) and suggesting some problems. That said, part of the answer to (1) maybe should be some guardrails such as "low external dependency requirements" and "problem statement should not be an obvious classic problem".
